### DISCUSSION TITLE
Installing software

### DISCUSSION THREAD

**Discussion by dirkhh** [MichaIng/DietPi#6584] [Q&A]
As I am working through adding a new software package (eventually there'll be a pull request)...
Where's the best place to ask questions about that?
And in case it's 'here' - are there some rules / guidelines that I haven't found yet? I'm looking at [how to add a software title](https://github.com/MichaIng/DietPi/wiki/How-to-add-a-new-software-title) but that doesn't really talk about the rules for the app itself. So, for example, are there rules for how to install systemd services? Are there rules about being able to claim a folder somewhere (e.g., a subfolder under `/opt`)
I'll stop here and see if I get some responses 🙃

**Comment by Joulinar** [comment]
Perfect place for such questions. @MichaIng as our developer will have answers on all your questions. 😉 Basically have a look to `dietpi-software`. There you will have a lot of examples on how things are done. However if you have improvement ideas, don't hesitate to share.

**Comment by dirkhh** [comment]
Yes, that's what I'm doing. Which is where the questions about `/opt` and about `systemd` services comes from.
I have posted a couple of improvement ideas elsewhere already (e.g. the discussion about a FAT partition to hold easy to edit files for the average user prior to first boot)

**Comment by Joulinar** [comment]
Well this I guess was explained by @MichaIng already why this is not that easy to achieve.

**Reply by dirkhh** [reply]
yes, putting `/boot` on FAT is challenging. But that's explicitly not what I'm suggesting :)
see https://github.com/MichaIng/DietPi/issues/6570#issuecomment-1697899163

**Reply by Joulinar** [reply]
Yes have seen your comment after I fired up this one 😉

**Comment by Joulinar** [comment]
Ah on thing in addition. We try to avoid Docker containers as install options and prefer native installations instead.

**Reply by dirkhh** [reply]
that's a noble goal - but in some scenarios it's not technically a sane approach. Let's say you create a feeder that wants to feed two dozen aggregators. Half of which are happy to use the same data format (that's one container), the other half insists on doing things their way, in some cases even with their own proprietary binaries (that's one container each). Trying to run them in the same system they keep stepping on top of each other.

**Comment by MichaIng** [comment]
Sounds great 👍. A look at the other software implementations should clarify most things:
- We prefer `/opt/name` as install/binary dir and `/mnt/dietpi_userdata/name` as data/config dir, if possible.
- We currently put systemd services into `/etc/systemd/system`, but actually I aim to change this to use `/usr/local/systemd/system` instead, so masks and `systemctl --full edit`.

**Reply by dirkhh** [reply]
With a mostly containerized complex system the concept of "binary" is a bit more complicated... everything except for the containers is actually source (Python, Shell). Right now (in its self contained form), all the settings, everything that possibly can lives under `/opt/adsb` - but I am happy to change this (as far as possible) to be under `/mnt/dietpi_userdata/adsb`. Which would leave the scripts and the Python app under /opt/adsb. That's a fair bit of code change, but I want to make sure that what I'm submitting doesn't get rejected right out of hand 🙃

As for the systemd stuff, I'll try to steal from, err, get inspired by other apps - I'm sure I can find some examples in the couple of hundred ones that are already there.

**Reply by MichaIng** [reply]
Sounds fine. Yes, I meant "install" or executable directory. There are more cases of software in source code form, when runtime systems are used etc. The ideas is that most userdata and configs can be saved/transferred to other systems if e.g. `/mnt/dietpi_userdata` is moved to an external drive, respectively to have (most of) them in a single place, while having dirs in `/opt` as something which can be usually just removed and redownloaded/installed as repair/reinstall/update step without any data loss. But there is software which expect their configs and store data in sub directories of the main install dir.

**Reply by dirkhh** [reply]
100% agree with the idea and my setup today is 100% NOT setup that way.
Which means that I need to re-shuffle the underlying software to adhere to this (smart) philosophy.

I expected this to be a learning experience and a significant lift. I appreciate your patience in answering my questions and helping guide me to the point where I can get this integrated.

**Reply by dirkhh** [reply]
BTW: this turned out to be "reasonably easy" and I submitted a PR to get more feedback on my approach.
